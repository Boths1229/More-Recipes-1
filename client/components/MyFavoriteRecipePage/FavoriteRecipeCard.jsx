import React, { Component, PropTypes } from 'react';
import { Link } from 'react-router';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Tooltip from 'react-tooltip';
import alert from 'sweetalert2';
import { favoriteARecipe } from '../../actionsCreator/recipes';


const propTypes = {
  actions: PropTypes.shape({
    favoriteARecipe: PropTypes.func
  }).isRequired,
  userId: PropTypes.number.isRequired,
  recipe: PropTypes.shape({
    id: PropTypes.number,
    image: PropTypes.string,
    recipeName: PropTypes.string
  }).isRequired,
  index: PropTypes.number.isRequired,
};

/**
 * @description A class to mount the favorite recipe card
 * @extends Component
 */
class FavoriteRecipeCard extends Component {
  /**
   * handles removing favorite recipe
   * @param {object} props
  */
  constructor(props) {
    super(props);

    this.favoriteRecipe = this.favoriteRecipe.bind(this);
  }

  /**
   * @description remove recipe  from favorites
   *
   * @memberof FavoriteRecipeCard
   *
   * @returns {void}
  */
  favoriteRecipe() {
    const {
      userId,
      recipe,
      index
    } = this.props;
    const data = {
      recipeId: recipe.id
    };
    alert({
      title: 'Are you sure?',
      text: "You won't be able to revert this!",
      type: 'warning',
      showCancelButton: true,
      confirmButtonColor: '#3085d6',
      cancelButtonColor: '#d33',
      confirmButtonText: 'Yes, delete it!',
      cancelButtonText: 'No, cancel!',
      confirmButtonClass: 'btn btn-danger',
      cancelButtonClass: 'btn btn-success  alert-button',
      buttonsStyling: false,
      reverseButtons: true
    }).then((result) => {
      if (result.value) {
        this.props.actions.favoriteARecipe(userId, data, index);
      } else if (
        result.dismiss === alert.DismissReason.cancel
      ) {
        alert(
          'Cancelled',
          'Your recipe is safe',
          'error'
        );
      }
    });
  }

  /**
   * @description renders the favorite recipe card
   *
   * @memberof FavoriteRecipeCard
   *
   * @returns {void} returns the components
  */
  render() {
    const {
      recipe
    } = this.props;
    return (
      <div className="col-md-4 card-space">
        <div className="card-deck">
          <article className="card recipe-card">
            <div className="my-recipe-img">
              <img
                src={recipe.image}
                alt="recipe"
                className="card-img-top img-fluid recipe-card-img"
              />
            </div>
            <div className="card-body">
              <h3 className="card-title">
                <Link to={`/recipes/${recipe.id}`} id="remove-link">
                  {recipe.recipeName}
                </Link>
              </h3>
            </div>
            <div >
              <div className="fav-btn">
                <button
                  data-tip
                  data-for="favorite"
                  className="btn btn-danger"
                  type="button"
                  onClick={this.favoriteRecipe}
                >
                  <i
                    className="fa fa-trash-o fa-2x"
                  />
                </button>
                <Tooltip
                  id="favorite"
                  place="top"
                  type="dark"
                >
                  <span>Remove from favorites</span>
                </Tooltip >
              </div>
            </div>
          </article>
        </div>
      </div>
    );
  }
}

/**
 * @description maps action to properties of FavoriteRecipeCard
 *
 * @param  {object} dispatch
 *
 * @returns {object} returns the action to be bind
 */
function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      favoriteARecipe
    }, dispatch)
  };
}

FavoriteRecipeCard.propTypes = propTypes;

export default connect(null, mapDispatchToProps)(FavoriteRecipeCard);
