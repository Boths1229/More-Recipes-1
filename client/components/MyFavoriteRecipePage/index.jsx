import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Spinner } from 'react-preloading-component';
import Header from '../common/AuthHeader';
import Footer from '../common/Footer';
import FavRecipeCard from './FavoriteRecipeCard';
import { getUserFavRecipes } from '../../actionsCreator/recipes';
import verifyUser from '../../util/Authentication';


const propTypes = {
  actions: PropTypes.shape({
    getUserFavRecipes: PropTypes.func
  }).isRequired,
  user: PropTypes.shape({
    id: PropTypes.number,
    username: PropTypes.string
  }).isRequired,
  favoriteRecipes: PropTypes.oneOfType([
    PropTypes.shape(),
    PropTypes.arrayOf(PropTypes.shape())
  ]).isRequired,
  isLoading: PropTypes.bool
};

const defaultProps = {
  isLoading: null
};

/**
 * @description A class to mount all components related to MyFavoriteRecipePage
 * @extends Component
 */
export class MyFavoriteRecipePage extends Component {
  /**
   * handles rendering favorite recipe
   * @param {object} props
  */
  constructor(props) {
    super(props);

    this.renderFavoriteRecipes = this.renderFavoriteRecipes.bind(this);
  }

  /**
   * @description performs an action right after the component mount
   *
   * @memberof MyFavoriteRecipePage
   *
   * @returns {void}
  */
  componentDidMount() {
    if (verifyUser() === true) {
      this.props.actions.getUserFavRecipes(this.props.user.id);
    }
  }

  /**
   * @memberof MyFavoriteRecipePage
   *
   * @returns {void}
  */
  renderFavoriteRecipes() {
    const {
      isLoading,
      favoriteRecipes,
      user
    } = this.props;
    if (isLoading) {
      return (
        <div style={{
          marginTop: '150px',
          marginLeft: 'auto',
          marginRight: 'auto',
        }}
        >
          <Spinner />
        </div>
      );
    }
    if (favoriteRecipes.message || favoriteRecipes.favorited.length === 0) {
      return (
        <h1 style={{ margin: '150px auto 0', width: '100%', textAlign: 'center' }}>
          You have no favorite recipes
        </h1>
      );
    }
    return favoriteRecipes.favorited.map(favorites => (
      <FavRecipeCard
        key={favorites.id}
        recipe={favorites.Recipe}
        favorite={this.favoriteRecipe}
        userId={user.id}
      />
    ));
  }

  /**
   * @description renders the components
   *
   * @memberof MyFavoriteRecipePage
   *
   * @returns {JSX} returns the components
  */
  render() {
    return (
      <div>
        <Header user={this.props.user} />
        <div className="container">
          <div className="row section-popular">
            <div className="col-sm-12">
              <h2 className="pt-3 headline centered">My Favorite Recipes</h2>
            </div>
          </div>
          <div className="row pt-4">
            {
              this.renderFavoriteRecipes()
            }
          </div>
        </div>
        <Footer />
      </div>
    );
  }
}

/**
 * @description maps state to properties of MyFavoriteRecipePage
 *
 * @param  {object} state
 *
 * @returns {object} returns the state to be mapped to props
 */
function mapStateToProps(state) {
  return {
    user: state.auth.user,
    favoriteRecipes: state.recipes.favoriteRecipes,
    isLoading: state.recipes.isLoading
  };
}

/**
 * @description maps action to properties of MyFavoriteRecipePage
 *
 * @param  {object} dispatch
 *
 * @returns {object} returns the action to be bind
 */
function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      getUserFavRecipes
    }, dispatch)
  };
}

MyFavoriteRecipePage.propTypes = propTypes;
MyFavoriteRecipePage.defaultProps = defaultProps;

export default connect(mapStateToProps, mapDispatchToProps)(MyFavoriteRecipePage);
