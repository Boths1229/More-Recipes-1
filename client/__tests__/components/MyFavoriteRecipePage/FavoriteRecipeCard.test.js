import React from 'react';
import { shallow } from 'enzyme';
import { FavoriteRecipeCard } from '../../../components/MyFavoriteRecipePage/FavoriteRecipeCard';
import MockData from '../../__mocks__/actions/recipes';

let props;

const setup = () => {
  props = {
    actions: {
      favoriteARecipe: () => {},
    },
    userId: 1,
    recipe: MockData.recipeData,
    index: 1,
  };

  return shallow(<FavoriteRecipeCard {...props} />);
};

describe('FavroiteRecipeCard', () => {
  it('should render correctly', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('img').length).toBe(1);
    expect(wrapper.find('button').length).toBe(1);
  });

  it('should remove recipe from favorites when the favoriteRecipe method is called', () => {
    const wrapper = setup();
    const action = wrapper.instance();
    const spy = jest.spyOn(action, 'favoriteRecipe');
    action.favoriteRecipe();
    expect(spy).toBeCalled();
  });
});
