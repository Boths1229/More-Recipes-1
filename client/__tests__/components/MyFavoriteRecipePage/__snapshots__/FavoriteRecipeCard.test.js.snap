// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavroiteRecipeCard should render correctly 1`] = `
<div
  className="col-md-4 card-space"
>
  <div
    className="card-deck"
  >
    <article
      className="card recipe-card"
    >
      <div
        className="my-recipe-img"
      >
        <img
          alt="recipe"
          className="card-img-top img-fluid recipe-card-img"
          src="https://firebasestorage.googleapis.com/v0/b/more-recipes-3be20.appspot.com/o/images%2F702446f3-a2c2-48c7-a9d1-af20be9281d7.jpeg?alt=media&token=1827adb2-4108-4939-9fe0-778e27dd5212"
        />
      </div>
      <div
        className="card-body"
      >
        <h3
          className="card-title"
        >
          <Link
            id="remove-link"
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/recipes/27"
          >
            new recipe
          </Link>
        </h3>
      </div>
      <div>
        <div
          className="fav-btn"
        >
          <button
            className="btn btn-danger"
            data-for="favorite"
            data-tip={true}
            onClick={[Function]}
            type="button"
          >
            <i
              className="fa fa-trash-o fa-2x"
            />
          </button>
          <ReactTooltip
            id="favorite"
            insecure={true}
            place="top"
            resizeHide={true}
            type="dark"
            wrapper="div"
          >
            <span>
              Remove from favorites
            </span>
          </ReactTooltip>
        </div>
      </div>
    </article>
  </div>
</div>
`;
